src/librustc_trans/back/bytecode.rs

```diff
@@ -0,0 +1,160 @@
+//! up too much space on disk.
+//!
+//! After compressing the bytecode we then have the rest of the format to
+//! basically deal with various bugs in various archive implementations. The
+//! format currently is:
+//!
+//!     RLIB LLVM-BYTECODE OBJECT LAYOUT
+//!     Version 2
+//!     Bytes    Data
+//!     0..10    "RUST_OBJECT" encoded in ASCII
+//!     11..14   format version as little-endian u32
+//!     15..19   the length of the module identifier string
+//!     20..n    the module identifier string
+//!     n..n+8   size in bytes of deflate compressed LLVM bitcode as
+//!              little-endian u64
+//!     n+9..    compressed LLVM bitcode
+//!     ?        maybe a byte to make this whole thing even length
+
+use std::io::{Read, Write};
+use std::ptr;
+use std::str;
+
+use flate2::Compression;
+use flate2::read::DeflateDecoder;
+use flate2::write::DeflateEncoder;
+
+// This is the "magic number" expected at the beginning of a LLVM bytecode
+// object in an rlib.
+pub const RLIB_BYTECODE_OBJECT_MAGIC: &'static [u8] = b"RUST_OBJECT";
+
+// The version number this compiler will write to bytecode objects in rlibs
+pub const RLIB_BYTECODE_OBJECT_VERSION: u8 = 2;
+
+pub const RLIB_BYTECODE_EXTENSION: &str = "bytecode.encoded";
+
+pub fn encode(identifier: &str, bytecode: &[u8]) -> Vec<u8> {
+    let mut encoded = Vec::new();
+
+    // Start off with the magic string
+    encoded.extend_from_slice(RLIB_BYTECODE_OBJECT_MAGIC);
+
+    // Next up is the version
+    encoded.extend_from_slice(&[RLIB_BYTECODE_OBJECT_VERSION, 0, 0, 0]);
+
+    // Next is the LLVM module identifier length + contents
+    let identifier_len = identifier.len();
+    encoded.extend_from_slice(&[
+        (identifier_len >>  0) as u8,
+        (identifier_len >>  8) as u8,
+        (identifier_len >> 16) as u8,
+        (identifier_len >> 24) as u8,
+    ]);
+    encoded.extend_from_slice(identifier.as_bytes());
+
+    // Next is the LLVM module deflate compressed, prefixed with its length. We
+    // don't know its length yet, so fill in 0s
+    let deflated_size_pos = encoded.len();
+    encoded.extend_from_slice(&[0, 0, 0, 0, 0, 0, 0, 0]);
+
+    let before = encoded.len();
+    DeflateEncoder::new(&mut encoded, Compression::Fast)
+        .write_all(bytecode)
+        .unwrap();
+    let after = encoded.len();
+
+    // Fill in the length we reserved space for before
+    let bytecode_len = after - before;
+    encoded[deflated_size_pos + 0] = (bytecode_len >>  0) as u8;
+    encoded[deflated_size_pos + 1] = (bytecode_len >>  8) as u8;
+    encoded[deflated_size_pos + 2] = (bytecode_len >> 16) as u8;
+    encoded[deflated_size_pos + 3] = (bytecode_len >> 24) as u8;
+    encoded[deflated_size_pos + 4] = (bytecode_len >> 32) as u8;
+    encoded[deflated_size_pos + 5] = (bytecode_len >> 40) as u8;
+    encoded[deflated_size_pos + 6] = (bytecode_len >> 48) as u8;
+    encoded[deflated_size_pos + 7] = (bytecode_len >> 56) as u8;
```

Comment:
`bytecode_len` is an `usize`. It should be cast to `u64`, or the last 4 statements should be `#[cfg]`'ed out.


Failed on `i686-gnu`:

```
[00:19:51] error: bitshift exceeds the type's number of bits
[00:19:51]   --> /checkout/src/librustc_trans/back/bytecode.rs:88:38
[00:19:51]    |
[00:19:51] 88 |     encoded[deflated_size_pos + 4] = (bytecode_len >> 32) as u8;
[00:19:51]    |                                      ^^^^^^^^^^^^^^^^^^^^
[00:19:51]    |
[00:19:51]    = note: #[deny(exceeding_bitshifts)] on by default
[00:19:51] 
[00:19:51] error: bitshift exceeds the type's number of bits
[00:19:51]   --> /checkout/src/librustc_trans/back/bytecode.rs:89:38
[00:19:51]    |
[00:19:51] 89 |     encoded[deflated_size_pos + 5] = (bytecode_len >> 40) as u8;
[00:19:51]    |                                      ^^^^^^^^^^^^^^^^^^^^
[00:19:51] 
[00:19:51] error: bitshift exceeds the type's number of bits
[00:19:51]   --> /checkout/src/librustc_trans/back/bytecode.rs:90:38
[00:19:51]    |
[00:19:51] 90 |     encoded[deflated_size_pos + 6] = (bytecode_len >> 48) as u8;
[00:19:51]    |                                      ^^^^^^^^^^^^^^^^^^^^
[00:19:51] 
[00:19:51] error: bitshift exceeds the type's number of bits
[00:19:51]   --> /checkout/src/librustc_trans/back/bytecode.rs:91:38
[00:19:51]    |
[00:19:51] 91 |     encoded[deflated_size_pos + 7] = (bytecode_len >> 56) as u8;
[00:19:51]    |                                      ^^^^^^^^^^^^^^^^^^^^
[00:19:51] 
[00:19:51] error: aborting due to 4 previous errors
[00:19:51] 
[00:19:51] error: Could not compile `rustc_trans`.
```